### ISSUE TITLE
Choosing fields during collection creation

### ISSUE THREAD

**Issue by bradfordcondon** [tripal/tripal#239]
### Problem 

Users don't know necessarily what fields they need to select to generate a given file.   For example @almasaeed2010 created thisFASTA file.  Edit: he did in fact select the sequence field, but it did not show up.  It does show up in the CSV.

```
>FRAEX38873_v2_000001970.1 FRAEX38873_v2_000001970.1  Sequence Coordinates: This feature is not located on any sequence.; Expression: 189; Time Accessioned: 2017-12-11 09:44:04.799761; Resource Type: mRNA; Organism: Fraxinus Excelsior; Sequence Checksum: b59e0ed1eaee7a3044794781680a6397; Sequence Length: 720; Time Last Modified: 2017-12-11 09:44:04.799761
>FRAEX38873_v2_000001350.1 FRAEX38873_v2_000001350.1  Sequence Coordinates: This feature is not located on any sequence.; Expression: 131; Time Accessioned: 2017-12-11 09:44:04.799761; Resource Type: mRNA; Organism: Fraxinus Excelsior; Sequence Checksum: 30f85b9c6ac69e8fdc1eee5cc7cd6e0a; Sequence Length: 2391; Time Last Modified: 2017-12-11 09:44:04.799761
```

### Discussion

On the one hand the FASTA downloader could handle the error checking (it does appear to so not sure why the above file was allowed) but to me, this raises a broader question:

**Should users tasked with selecting fields when creating collections?**  Even if we marked which fields would be required for which file types, is this something users should have to do?  It's pretty cumbersome, a good deal of the fields are confusing, there are a lot of them, and you can fail to include the fields you would want.  

I have two ideas...

* An admin page, where admins configure which fields for each entity are included in a collection.
* Fields are not selected at all.  Collections are lists of entities.  Instead, the specific instance of the downloader class is responsible for collecting the relevant fields (and perhaps getting user input on which fields to include)

**Comment by almasaeed2010** [comment]
Hi all,

We also noticed that when a collection job fails, there is no way for the end-user to find out. It currently continues to say `waiting...` even after the job quits.

Thanks!

**Comment by laceysanderson** [comment]
I'm also tempted to add the "bug" tag for

> He did in fact select the sequence field, but it did not show up. It does show up in the CSV.

and

> We also noticed that when a collection job fails, there is no way for the end-user to find out. It currently continues to say waiting... even after the job quits.

but I figure we'll wait and see what the discussion brings up.

**Comment by laceysanderson** [comment]
My thoughts on this are:
- I don't like having to select fields. It can be cumbersome for large lists...
- I don't want to have to administer what fields should go in a collection...
     - arguably not all admin know what fields are needed
     - this adds work for administrators and this makes Tripal harder to admin
- I DO like the idea of a collection simply being a collection of entities and downloaders pick which fields they need
- I imagine this feature (choosing fields) was originally added with CSV/TSV files in mind since including every field makes for a Lot of columns which the user might not even need/want.
- Perhaps we need a hook on a generate files page so each file format can collect more information from the user before generating a file? This would allow CSV/TSV file formats to let the users select fields but not require it for FASTA

**Comment by bradfordcondon** [comment]
really really good points.  I think you're especially on point noting that choosing fields probably makes sense with CSV/TSV.... **but its probably too confusing** for your average end user because so many fields attach!

What if instead we offered several flavor of "fixed" CSV/TSV downloaders?  I like users and admins not having to worry about fields **at all** but still making it easy for developers to customize the CSV/TSV file downloads.

**Comment by laceysanderson** [comment]
I'm intrigued... We would want a combination of generic and specific "Fixed" TSV/CSV download profiles. Maybe a hook to allow developers to specify a download type name + field array?

**Comment by bradfordcondon** [comment]
>Maybe a hook to allow developers to specify a download type name + field array?

  My first thought was handling it by having TSV/CSV downloaders being an *abstract class*, and these classes being extended by specifying the bundle type and field, which amounts to the same thing.  Hooks are maybe more intuitive to Drupal users, extending the class to non-Drupal devs.

**Comment by laceysanderson** [comment]
Since Tripal is moving towards classes (as is the whole Drupal world), your approach @bradfordcondon is likely the better one :-) @spficklin what are your thoughts on this?

**Comment by spficklin** [comment]
is it really a burden for users to select fields?  BioMart allows this.  By default, when creating collections using Drupal Views, the fields in the view are pre-selected and the user only need add fields if they want more details.  Also, the supported file type for each field is listed next to the field.  See screenshot:

![image](https://user-images.githubusercontent.com/1719352/48453335-ceaea580-e767-11e8-8c5b-dbc6a4fd2887.png)

But, I agree, this can be problematic if someone wants to export a GFF file. You have to specify all of the require fields or you won't be able to generate such a file correclty.  The GFF downloader is currently not finished for that reason.  Or, even for the FASTA file if you don't include the feature name, or uniquename the FASTA file will look odd.  

@laceysanderson the downloaders are already in Class form so we should be able to add a new member variable that could support this.  Perhaps a supported_entity_types and required fields.  If the data collection has entities of the supported type and the user wants a GFF file then we can pre-select all of the required fields.

**Comment by bradfordcondon** [comment]
>By default, when creating collections using Drupal Views, the fields in the view are pre-selected and the user only need add fields if they want more details. Also, the supported file type for each field is listed next to the field.

I think the elasticsearch collection form could be improved in this regard, i don't think we displayed the file information OR the field information.  Still, you can have a LOT of fields.  And an end user is likely to not know what a large number of them are.   But more importantly

>is it really a burden for users to select fields?

the user I have in mind is *not* super tech literate, so yes it is a pretty big deal imo to bring up a UI of all those fields, rather than a a couple of canned choice.  @mestato care to weigh in? Maybe I'm making a bigger deal of this issue than i need to.

**Comment by spficklin** [comment]
So, the fields do have descriptions for what they are and I don't think it requires someone tech savy.  But I do recognize that an unfamiliar interface may not be intuitive to a new user.   To play devil's advocate, is the extra code development needed to help with this worth it, or would a link to a tutorial on the page be sufficient for now?

**Comment by mestato** [comment]
I am completely on board with the Bradford/Lacey solution. Elasticsearch results to collections functionality doesn't display the field descriptions yet, which I admit would definitely help a lot. But choosing fields when building a file and only if necessary still makes the most sense to me.

As to file formats where the fields must be selected correctly up front - gff is one, vcf is another.  From the perspective of creating files to run in a Galaxy workflow, these are the types of format that will be needed, and users will probably not have any idea what fields are needed to create them.

Here's my understanding of what we are talking about. Lets assume a user has collections of entities (no fields at this point). Then at the stage of building a file, they choose a file format - if the site supports GFF for this data type, the user can select it, the code will go get the needed fields and build a gff. Same for fasta. But the user can also choose CSV or TSV. Then at that point, the user can select the fields for that particular file.  This seems to me the most robust solution for supporting standard file formats and not frustrating users while also having a lot of flexibility in what can be downloaded.

**Comment by laceysanderson** [comment]
I actually had to remove collections on KnowPulse due to too many user complaints regarding the complexity of choosing fields. So yes, I feel the extra development time is worth it 🤷‍♀️ 

When a user is just trying to save their results for later, they don't want to have to worry about which information they may need and they seem to be quite overwhelmed with the list. Moving the field selection to the file generation step and only for file formats where it makes sense will make the choices easier since it's at the time when they're actually downloading the results.

**Comment by spficklin** [comment]
I think this is easy enough to fix.  I'm going to be optimistic and assume we can get a fix for the 3.2 release on this.

**Comment by laceysanderson** [comment]
Just to weigh in on this to get the discussion going again. 

- While I feel the development time is worth it, I am not using collections and thus cannot prioritize this development. 
- @mestato how are you feeling on this functionality and may you have a developer with time to contribute towards a fix?
- @spficklin could you describe what you solution would look like so someone else could potentially pick it up?
